### PR TITLE
02 dynamic replaceable content

### PR DESCRIPTION
**Pull request by kitsao**
#2

### CODE REVIEW

File: dynamic-rapidpro-workspace-settings/.eslintrc [hunk 1]
```diff
@@ -1,6 +1,6 @@
 {
   "parserOptions": {
-    "ecmaVersion": 2017
+    "ecmaVersion": 9
```
Comment (derickl) [inline dynamic-rapidpro-workspace-settings/.eslintrc:3]: You can use either the numerical versioning or by year. Either option works
https://eslint.org/docs/latest/user-guide/configuring/language-options

File: dynamic-rapidpro-workspace-settings/utils.js [hunk 2]
```diff
@@ -51,41 +48,57 @@ const getInputs = (core) => {
...
+    if (fs.existsSync(flowsFile)) {
+      const rawFlowsContent = fs.readFileSync(`${codeRepository}/${flowsFile}`, 'utf8');
+      flowsContent = await getReplacedContent(rawFlowsContent, flatten(secrets));
+    }
 
-    await axios.put(url.href, `"${secrets.rp_api_token}"`);    
+    await axios.put(url.href, `"${secrets.rp_api_token}"`);
     fs.writeFileSync(`${codeRepository}/${settingsFile}`, settings);
-    fs.writeFileSync(`${codeRepository}/${flowsFile}`, getFormattedFlows(secrets.rp_flows));
+    fs.writeFileSync(`${codeRepository}/${flowsFile}`, flowsContent);
```
Comment (derickl) [inline dynamic-rapidpro-workspace-settings/utils.js]: Is it expected that we may have write back a null value?

File: dynamic-rapidpro-workspace-settings/utils.js [hunk 3]
```diff
@@ -51,41 +48,57 @@ const getInputs = (core) => {
...
     core.info(`Successful`);
   } catch (error) {
     core.setFailed(error.message);
   }
 };
 
+const flatten = obj => {
```
Comment (derickl) [inline dynamic-rapidpro-workspace-settings/utils.js]: consider giving it a name that makes it easier to understand it's function. e.g flattenObject.

You should probably throw if obj is not an object

File: dynamic-rapidpro-workspace-settings/test/index.spec.js [hunk 4]
```diff
@@ -90,17 +90,30 @@ describe(`rapidpro action test suite`, () => {
...
     expect(isValid).to.be.false;
   });
 
   it(`should be true if valid flows data is given`, async () => {
-    const isValid = utils.isValidFlows(secrets.rp_flows);
+    const isValid = utils.isValidObject(secrets.outbound_mapping_exprs);
     expect(isValid).to.be.true;
   });
+
+  it(`should return a flat object if nested is given`, async () => {
```
Comment (derickl) [inline dynamic-rapidpro-workspace-settings/test/index.spec.js]:
```suggestion
  it(`should return a flat object if nested object is given`, async () => {
```

File: dynamic-rapidpro-workspace-settings/test/index.spec.js [hunk 5]
```diff
@@ -90,17 +90,30 @@ describe(`rapidpro action test suite`, () => {
   });
 
   it(`should be false if invalid flows data is given`, async () => {
-    let isValid = utils.isValidFlows({});
+    let isValid = utils.isValidObject({});
     expect(isValid).to.be.false;
```
Comment (derickl) [inline dynamic-rapidpro-workspace-settings/test/index.spec.js]: expect(utils.isValidObject({})).to.be.false;

You don't need the intermediate step to hold a value

Comment (kitsao) [reply dynamic-rapidpro-workspace-settings/test/index.spec.js]: True.

File: dynamic-rapidpro-workspace-settings/utils.js (hunk 2 above)
Comment (kitsao) [reply dynamic-rapidpro-workspace-settings/utils.js]: Oops! No.

File: dynamic-rapidpro-workspace-settings/.eslintrc (hunk 1 above)
Comment (kitsao) [reply dynamic-rapidpro-workspace-settings/.eslintrc:3]: Numerical versioning is cleaner.

File: dynamic-rapidpro-workspace-settings/test/index.spec.js [hunk 6]
```diff
@@ -90,16 +90,26 @@ describe(`rapidpro action test suite`, () => {
...
   });
 
   it(`should be true if valid flows data is given`, async () => {
-    const isValid = utils.isValidFlows(secrets.rp_flows);
-    expect(isValid).to.be.true;
+    expect(utils.isValidObject(secrets.outbound_mapping_exprs)).to.be.true;
+  });
+
+  it(`should return a flat object if nested object is given`, async () => {
+    const n = Object.keys(secrets).length;
```
Comment (derickl) [inline dynamic-rapidpro-workspace-settings/test/index.spec.js:103]: single character variable names are harder grok. consider renaming to communicate intent. e.g `originalLength`

File: dynamic-rapidpro-workspace-settings/utils.js [hunk 7]
```diff
@@ -51,41 +48,62 @@ const getInputs = (core) => {
...
+    if (fs.existsSync(flowsFile)) {
+      const rawFlowsContent = fs.readFileSync(`${codeRepository}/${flowsFile}`, 'utf8');
+      flowsContent = await getReplacedContent(rawFlowsContent, flattenedSecrets);
+    }
 
-    await axios.put(url.href, `"${secrets.rp_api_token}"`);    
+    await axios.put(url.href, `"${secrets.rp_api_token}"`);
     fs.writeFileSync(`${codeRepository}/${settingsFile}`, settings);
-    fs.writeFileSync(`${codeRepository}/${flowsFile}`, getFormattedFlows(secrets.rp_flows));
+    if(flowsContent !== null){
```
Comment (derickl) [inline dynamic-rapidpro-workspace-settings/utils.js:77]: could be simplified to `if (!flowsContent)`

File: dynamic-rapidpro-workspace-settings/test/index.spec.js [hunk 8]
```diff
@@ -90,16 +90,26 @@ describe(`rapidpro action test suite`, () => {
   });
 
   it(`should be false if invalid flows data is given`, async () => {
-    let isValid = utils.isValidFlows({});
-    expect(isValid).to.be.false;
-
-    isValid = utils.isValidFlows(0);
-    expect(isValid).to.be.false;
+    expect( utils.isValidObject({})).to.be.false;
```
Comment (derickl) [inline dynamic-rapidpro-workspace-settings/test/index.spec.js:93]: delete the space after the parenthesis

File: dynamic-rapidpro-workspace-settings/utils.js [hunk 9]
```diff
@@ -51,41 +48,62 @@ const getInputs = (core) => {
...
 
 const run = async (githubWorkspacePath, params, fs) => {
   try {
     if (!githubWorkspacePath) {
       throw new Error(`GITHUB_WORKSPACE not defined`);
     }
     const secrets = getInputs(params);
-    if (!isValidFlows(secrets.rp_flows)) {
-      throw new Error(`Invalid flows data`);
+    const flattenedSecrets = flattenObject(secrets);
```
Comment (derickl) [inline dynamic-rapidpro-workspace-settings/utils.js:59]: Do you need to  reference both `secrets` and `flattenedSecrets`? Would working with flattenedSecrets be sufficient?

**Review by derickl — APPROVED** [review]
LGTM

Left some feedback inline

**Pull request merged**